### ISSUE TITLE
Support multiple versions of docs.

### ISSUE THREAD

**Issue by jimporter** [mkdocs/mkdocs#193]
I'd like to be able to host different docs for different versions of my code. I know there are services like ReadTheDocs that handle this for you, but it would be nice to be able to deploy all the versions to GitHub instead. This would make the gh-deploy script a bit more complex (and there'd need to be some way of switching versions when browsing), but I think it would be generally useful.

Maybe there's an argument to be made that GitHub Pages are for hosting the in-development version of your docs, and places like ReadTheDocs are for released versions; I'm not sure. My instinct is that all the versions of the docs should be hosted in the same place, though.

Any thoughts about this?

**Comment by d0ugal** [comment]
I've started making a tool that makes this possible. It is a very rough and early alpha, but it is a good start. I just need to add a version switcher ala ReadTheDocs or the django docs and it should work well.

It is somewhat opinionated, as it assumes you want to build all tags and that you are using git. This is why I am building it externally for now, this could be changed but I didn't want to get blocked from experimentation.


http://d0ugal.github.io/mkdocs-versioned/

**Comment by tomchristie** [comment]
@d0ugal Sounds neat!

**Comment by ePirat** [comment]
I agree with @d0ugal and think that the versioning of docs should be based on git tags as most docs are under git version control anyway. There are some important things to consider, though.
Given that older versions might have older configs, one possible approach would be that only the config of latest should be used to build all older docs (one corner case would be that older versions could require special markdown extensions that were disabled in newer configs, so this needs to be thought through a bit how it can be solved).
A version switcher like readthedocs has would be great, although this would make it mandatory to regenerate all old docs as well, when publishing new ones. So this would be another important thing to discuss. It would as well have the benefit that older docs can include a warning that these are for older versions and deprecated. (There are actually a lot of people reading old docs by accident)

To sum up the important points I would like to suggest for discussion:

- Use git tags for versioning?
- Only use current (latest) config?
- Have a version switcher in the docs/deprecation warnings?

**Comment by iwootten** [comment]
Hey @dougal - that looks great, but I can't see any files listed in pypi? https://pypi.python.org/pypi/mkdocs-versioned/0.1.0.dev0 I get an error about 'No distributions at all found for mkdocs-versioned==0.1.0.dev0'. I attempted installing from your setup.py script in the repo (which seems to have worked), but can't see it on the commandline?

**Comment by iwootten** [comment]
Scratch that, pip -e mkdocs-versioned seems to have done the job. Would be good to see it on pypi though.

**Comment by d0ugal** [comment]
@iwootten Hmm, yeah. It shouldn't really be on PyPI at all right now as it only vaguely works.

**Comment by iwootten** [comment]
Ah, sorry didn't realise your documentation was ahead of itself right now.

**Comment by d0ugal** [comment]
@iwootten no worries, there is a fairly big fat warning near the top of the documentation :) http://d0ugal.github.io/mkdocs-versioned/

**Comment by robodude666** [comment]
@d0ugal Are you intending to use `mkdocs-versioned` as a basis for a plugin that will help make versioning possible, or is it intended to be merged into `mkdocs`?

Personally, I think it should be handled via an external plugin as there are many different ways of versioning docs. I also don't see the value in rebuilding all of the docs, and especially not all tags.

Python's doc versioning is handled by a javascript file that adds a dropdown menu w/ the latest versions. This file only needs to be updated to include the latest a.b.C number or when an a.B.c release is deleted.

**Comment by d0ugal** [comment]
Yeah, I plan for it to be handled externally. I've said it a few times, but I forget where, so to quickly reiterate: that repo implements a pattern very similar to how ReadTheDocs versions Sphinx projects and it is also the pattern I want to use for MkDocs itself.

**Comment by d0ugal** [comment]
And to be clear, at the moment it uses private API's, I'd expect it to use a plugin API if the idea is to survive. However, that is dependant on #206

**Comment by waylan** [comment]
>  I also don't see the value in rebuilding all of the docs, and especially not all tags.

This depends on how your docs are deployed. For example, when deploying to PyPI, you are required to upload a zip file which contains all of your site's files. The contents of that file will replace everything from the previous upload. No other options are provided.

Of course, if I have a local cache of the previous builds, I could rely on that and only build the most recent version before zipping up the entire thing (all versions), but I don't always work from the same machine so my local builds could get out-of-sync and local builds sometimes need to be cleaned out/deleted. So yes, we definitely need to be able to build all tags every time. Of course, the option to restrict builds to certain tags/versions would be useful as well.

**Comment by waylan** [comment]
One thing I haven't seen mentioned here yet is how to address a project which was not using MkDocs in older versions/tags. For example, my project has been around for many years before MkDocs with many versions tagged in the git repo. Interestingly, most of those tags also have docs and a one-off command which will build those docs. I realize I could create by own script which steps through all tags and build the docs, but it would be nice to be able to use a single tool for pre-MkDocs and post-MkDocs versions. 

I'm thinking perhaps a optional flag to set a cutoff tag and the option to define a build command for anything before that cutoff. Any tags before the cutoff would use the provided pre-cutoff command and any tags after the cutoff would use MkDocs.

Another issue is that prior to version 1.0, MkDocs is unstable and some older versions of some projects will be using what are now invalid configs or command line flags which are no longer supported by the CLI. Of course, the tagged version of some third-party project will not match MkDocs' versions, so it would be impossible for the script to determent which versions of the docs it can build.

And you can't require that some config file be added (or add a setting to the MkDocs config), as these tags are in past history of peoples repos. No one should be required to go through their projects and rewrite their history to add/change the MkDocs config for older versions. Besides, that would break forks/clones of the repos and is generally considered bad practice for version control systems.

I suppose a config in the current version could define various changes throughout history, and then the script could load that history and as it looped through tags it could adapt its behavior. But that sounds overly complex to me and probably not something anyone would want to maintain. That said, this is a real problem that exits in the real world. It wouldn't be right to completely ignore it.

**Comment by robodude666** [comment]
@waylan I see; wasn't aware of PyPi's requirements.

Regarding supporting non-mkdocs and pre-1.0 mkdocs versions, I don't think any of that should be handled by such a versioning plugin because you run the risk of adding too much bloat in order to support every possible scenario. Instead, something like this should be left to the project's developers to solve and could easily be handled via tox (or manually through virtualenv), allowing you to setup different environments and manually including commands to build from specific tags.

**Comment by waylan** [comment]
> Regarding supporting non-mkdocs and pre-1.0 mkdocs versions, I don't think any of that should be handled by such a versioning plugin because you run the risk of adding too much bloat in order to support every possible scenario. 

I don't disagree with you. I'm just saying that this is a real situation and it would be nice if it was acknowledged. Providing a way to explicitly specify which tags to built should work fine. Whether that is best achieved by specifying each tag individually, providing a cutoff (everything after a given tag), or providing start/stop tags (everything between), I don't know. And do you specify those tags via a config file, command line options, or both?

Personally, I would think the cutoff would be the simplest to implement. A single command line option could set the cutoff and every tag greater than or equal to the given tag would get built. IIRC, Distutils/Setuptools has some code for working with semantically versioned numbers (`x.x.x`) that could be used to sort/splice the list of tags. I would be okay with the restriction that the cutoff option only works with semantic versioning.

However, if someone wants to support multiple versions of MkDocs though-out their history they would need create their own script which created virtualenvs for each MkDocs version and build the docs for that version separately. As you suggest, something like Tox might work well for that. I suppose there would be no reason for this tool to support that scenario, making the start/stop option unnecessary.

**Comment by d0ugal** [comment]
What I did (or was going to do, I forget how far I got!), which was fairly easy, was build every tag with a mkdocs.yml. If the config didn't exist it was skipped and logged at DEBUG level.

Some tags failed because the config was really out of date and it didn't work with MkDocs anymore - these were logged with a warning.

**Comment by waylan** [comment]
I suppose that works as well.

**Comment by d0ugal** [comment]
It is reasonably good as a simple approach, but it does have a number of short comings like old tags start failing as MkDocs progresses - although, we have been quite good so far, so I don't expect many breakages and plan to avoid them as much as possible.

**Comment by jimporter** [comment]
> So yes, we definitely need to be able to build all tags every time. Of course, the option to restrict builds to certain tags/versions would be useful as well.

It seems to me that this is a sign that the built docs should be under (some form of) version control.

I look at documentation as a part of packaging a release; building the HTML version of a release's docs isn't really that different from building a source (or binary) tarball of a release. In fact, you might want to package the HTMLized docs inside your tarballs.

Following this view, the docs should be built exactly once, when you create the release. If you're hosting your docs on GitHub pages, you might want to just add the docs in the appropriate subfolder and push it to your gh-pages branch. If you're hosting your docs on PyPI, you might have a similar branch of your repository that you can use to manage the sum of all versions of your docs.

This makes it easier to handle docs that change format over time or even just use an old, incompatible version of a single doc generator. It also makes a promise to the user that the docs for a release won't change. This is similar to the idea of "semantic versions": under semantic versioning, you wouldn't re-release your code with the same version, even if all you did was fix a tiny bug. I think there's even an argument for making a new point-release just for documentation fixes: it's a way of letting users know that there's new stuff. (I doubt most projects will be this strict about versioning, though.)

**Comment by victorcoder** [comment]
Upvoting this :+1:

**Comment by fleed** [comment]
+1 for this feature

**Comment by schollz** [comment]
+1

**Comment by thenitai** [comment]
:+1:

**Comment by andrejkosyak** [comment]
+1

**Comment by huegelc** [comment]
Looking forward to see this feature too

**Comment by rlove** [comment]
Not sure I like changing the config file to control the output directory as it's not as dynamic and requires code to modify the config file before running mkdocs given the following scenario.


It would be nice if the gh-deploy deployment functionality could be changed to have optional subdirectory

````
mkdocs gh-deploy --clean --dir REPO_DIR
`````
The primary goal is to have multiple sets of documentation hosted under gh-pages.

| Directory | Purpose |
| ---------- | --------- |
| /current   | Latest commit Documentation |
| /release/v1.0.0.1 | Version 1.0.0.1 Documentation  |
| /release/vX.X.X.X | Version X.X.X.X Documentation |

So I would be able to call it like this to deploy to the `/release/vX.X.X.X` directory

````
mkdocs gh-deploy --clean --dir /release/vX.X.X.X
`````

**Comment by jschristie** [comment]
hey all, 

I am not using readthedocs.org or gh-pages, rather i am using a custom theme from material for mkdocs, and i have further customized it. Is there a way to make the versioning of docs using branches, with different mkdocs.yml

**Comment by waylan** [comment]
@jschristie the most recent work done on this issue is mentioned in [this comment][1]. No my knowledge, no work has been done on this since as we have had other priorities. Of course, this is all open source so you are free to experiment. If you make any progress, please pass it along.

[1]: https://github.com/mkdocs/mkdocs/issues/193#issuecomment-104901162

**Comment by xiaods** [comment]
looking forward for it.

**Comment by Himura2la** [comment]
+1 we need this

**Comment by EvanHerman** [comment]
+1 This would be nice to have 👍

**Comment by jimporter** [comment]
It's still very basic, but I've written an external to do this the way I suggested above (deploy the *current version* of your docs to a subdirectory on your `gh-pages` branch). I wouldn't recommend using it for anything mission-critical just yet, but it should work for basic scenarios. If you're interested, here it is: <https://github.com/jimporter/mkultra>.

I'd be happy to merge this in to MkDocs proper or maintain it as a separate tool; I'm not sure if there's any kind of consensus yet about how an "official" multi-version feature would work (@d0ugal's version works quite differently from mine).

**Comment by waylan** [comment]
@jimporter very cool. Of course, as you know, that's not a general purpose solution as it requires a git based deploy. However, I find our set of commands to interact very interesting.

**Comment by jimporter** [comment]
@waylan That's true; however, the only thing you'd really need for deploying to a local directory would be a way to manage the `versions.json` file. Everything else could really be handled with stock `MkDocs` plus the extra JS I made for the version selector. I'll probably wait on doing that though, since it's not essential for my use case, but it might be a good idea if there's demand for it.

**Comment by jimporter** [comment]
For those who are interested, I finally released mkultra on PyPI: https://pypi.python.org/pypi/mkultra

It should work out of the box with the built-in themes plus the bootswatch ones. Feel free to file issues and such if something's broken!

**Comment by jimporter** [comment]
> For those who are interested, I finally released the tool I wrote for this on PyPI: https://pypi.python.org/pypi/mike

Just a quick update: this is now up to v0.3.1, and it covers all the use cases I care about, so hopefully it will work for others as well! Per the discussion above, I don't know if `mike` is something that could ever get merged into MkDocs proper, but I'm happy to discuss how we'd go about doing that if there's interest.

**Comment by ldez** [comment]
Hello, we (@mmatur and me) have created a small tool https://github.com/containous/structor.

Use a different approach than [mike](https://pypi.python.org/pypi/mike) (which is a great tool):
- the versions are build in a Docker Image to avoid issues with versions of libraries or Python.
- works with all themes and each doc version can use different theme.
- one binary to rule them all

:smile_cat:

**Comment by cirosantilli** [comment]
I have seen the lite, use single page README.adoc: http://www.cirosantilli.com/markdown-style-guide/#use-asciidoc

**Comment by Niondir** [comment]
+1 for integrating this feature into mkdocs

**Comment by bilal-fazlani** [comment]
+1 for this feature

**Comment by snim2** [comment]
+1

**Comment by pblitz** [comment]
+1

**Comment by jimporter** [comment]
@waylan Do you have any thoughts on what we should do about this issue? I know there are a few different solutions out there for handling versioning (my `mike` project among them), but maybe it would make sense to converge on one of them. That could mean directly integrating multi-version handling into MkDocs, or maybe the MkDocs org just helps manage on of the existing solutions out there.

This would be particularly helpful for me and the `mike` project, since it's never really been my main open source focus, and with `mkdocs-material` having built-in support for `mike`, I fear that I won't have enough free time to devote to support/maintenance on my own. It'd be nice to have a long-term home for it where people can help out, even if I remain the primary maintainer for a while.

**Comment by waylan** [comment]
The problem is that I have less time to devote to MkDocs these days. I can barely keep up with current maintenance now. Adding this feature would be  larger undertaking than I am interested in taking. As far as I am aware, the existing solutions work fine therefore, my thinking is that they can continue to work as third party plugins.

**Comment by jimporter** [comment]
@waylan Yeah, that's where I feel like I'm heading towards with `mike` as well. I don't know if it'd be possible to get more contributors out there able to help maintain these projects; obviously, a lot of people use MkDocs, and perhaps a few of them would be able to help with maintenance. I'm happy to help out where I can, but my available time is pretty limited too.

Do you have any plans of trying to build a larger core of maintainers? That would hopefully take some of the load off and put MkDocs (and hopefully `mike` too, one day) on more stable footing. I'm not sure of the best way to go about doing that though; maybe an announcement in the README to an issue/discussion would attract some potential helpers?

**Comment by smoyergh** [comment]
Just for completeness, here is the approach that Docusaurus is taking to versioning: https://v2.docusaurus.io/docs/versioning

**Comment by BetterThanReal** [comment]
I'm not convinced that a lot of settings within `mkdocs.yml` actually work as described for many scenarios - for example, I'm confused why `site_url` lets me specify a URL that shows up in `sitemap.xml` and `<link rel="canonical">`, but yet are completely broken on the rendered site when I push via `mkdocs gh-deploy`.

If these settings actually worked as intended, they could be augmented with a few additional settings that could go a long way towards increasing the flexibility for various deployment scenarios.

Here's a solution that I would be happy to use for GitHub Pages:

1. Allow me to specify a subdirectory prefix, either in `mkdocs.yml` or via the command line, that would be prepended to the generated site content directory hierarchy, both on the file system, and the generated URLs.  For example, if I'm working on version 1.1.10 of my API, I would specify a subdirectory prefix of `1.1.10`, and I would expect to see the generated documents be hosted at `https://SITE/1.1.10/`
2. Push that subdirectory alone to the remote repository, merging it in without deleting the existing contents.

Problem solved.

**Comment by pawamoy** [comment]
This has been implemented by @jimporter: [mike](https://github.com/jimporter/mike). Since there's a solution to this in the form of a third-party, we decided to close this issue as out of scope for MkDocs' core. Feel free to comment further. We'd be especially interested in knowing if and how MkDocs core could help such versioning plugins.
